File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
   match arg with
   | "+" -> CalculatorOperation.Plus
   | "-" -> CalculatorOperation.Minus
   | "*" -> CalculatorOperation.Multiply
   | "/" -> CalculatorOperation.Divide
   | _ ->  ArgumentException() |> raise
  
let parseToDouble (str:string) =
    match Double.TryParse str with
    | true,double -> Some double
    | _ -> ArgumentException() |> raise

let parseCalcArguments(args : string[]) =
     if args = null || not(isArgLengthSupported args) then ArgumentException() |> raise
     else
     {
        arg1 = (parseToDouble (args[0])).Value
        operation = parseOperation (args[1])
        arg2 = ( parseToDouble (args[2])).Value
     }
    
    

File: Hw8/Calculator/Parser.cs
using System.Globalization;

namespace Hw8.Calculator
{
    public static class Parser
    {
        public static void ParseCalcArguments(string[] arr, out double val1, out Operation opr, out double val2  )
        {
            if (Double.TryParse(arr[0], NumberStyles.AllowDecimalPoint,
            CultureInfo.InvariantCulture, out val1) == false) { throw new ArgumentException(Messages.InvalidNumberMessage); }

            if (Double.TryParse(arr[2], NumberStyles.AllowDecimalPoint,
            CultureInfo.InvariantCulture, out val2) == false) { throw new ArgumentException(Messages.InvalidNumberMessage); }

            opr = ParseOperation(arr[1]);

            if (opr == Operation.Invalid) { throw new InvalidOperationException(Messages.InvalidOperationMessage); }

        }

        private static Operation ParseOperation(string str)
        {            
            switch (str)
            {
                case "Plus":
                    return Operation.Plus;
                    break;

                case "Minus":
                    return Operation.Minus;
                    break;

                case "Multiply":
                    return Operation.Multiply;
                    break;

                case "Divide":
                    return Operation.Divide;
                    break;

                default:
                    return Operation.Invalid;
                    break;
            }
        }
    }
}
